Feature/vlan

minsikl opened the pull request:
Provides softlayer_network_vlan resource. 3 files are added as follows:
1. docs/softlayer_network_vlan.md
2. softlayer/resource_softlayer_network_vlan.go
3. softlayer/resource_softlayer_network_vlan_test.go

vendor directory has been updated to provide a new helper function hardware.GetRouterByName()

Commit:
Initial commit of vlan resource

srinivas-dandibhotla committed:
Initial commit of vlan resource in feature/vlan branch

srinivas-dandibhotla committed:
Initial commit of vlan resource in the new feature/vlan branch

srinivas-dandibhotla committed:
Updated schema along with read and delete functions

minsikl committed:
Update schema to support price, child_resource_count, and subnet in network_vlan

srinivas-dandibhotla committed:
Updated schema and added conditional checks for type property

srinivas-dandibhotla committed:
Initial commit of acceptance test for network vlan resource

srinivasdandibhotla committed:
Updated documentation for network_vlan resource

srinivasdandibhotla committed:
Proper formatting of softlayer_network_vlan header

minsikl committed:
Implement create and update schema in network_vlan

minsikl committed:
Resolve merge conflict for network_vlan

minsikl committed:
Resolve merge conflic for README.md

minsikl committed:
govendor update

minsikl committed:
Add a documentation for network_vlan.

minsikl committed:
Update softlayer_network_vlan.md

minsikl committed:
Update VLAN package finding logic in network_vlan

ADDITIONAL_SERVICES and ADDITIONAL_SERVICES_NETWORK_VLAN packages provide
VLAN and Subnet items. ADDITIONAL_SERVICES provides /27, /28, and /29 subnet
items and ADDITIONAL_SERVICES_NETWORK_VLAN package provides /26, /28, and /29
subnet.
Create function tries to find items with package ADDITIONAL_SERVICES_NETWORK_VLAN
at first. If it fails, it will search items with ADDITIONAL_SERVICES.

minsikl committed:
Update test scenarios in vlan_test.

minsikl committed:
Merge branch 'feature/vlan' of github.com:softlayer/terraform-provider-softlayer into feature/vlan

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	vlanType := d.Get("type").(string)
+	if vlanType != "PRIVATE" && vlanType != "PUBLIC" {
+		return fmt.Errorf("Error creating network vlan: vlanType should be either 'PRIVATE' or 'PUBLIC'")
+	}
```
I think this is better done as a validate function on the schema.

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+			"vlanNumber," +
+			"billingItem[recurringFee]," +
+			"guestNetworkComponentCount," +
+			"subnets[networkIdentifier,cidr,subnetType]",
```
Performance nit. This is better to have as some constant globally in the package.

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+		d.Set("name", *vlan.Name)
+	} else {
+		d.Set("name", "")
+	}
```
`d.set("name", sl.Get(vlan.Name, ""))`

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+
+	for _, elem := range vlan.Subnets {
+		subnet := make(map[string]interface{})
+		subnet["subnet"] = *elem.NetworkIdentifier + "/" + strconv.Itoa(*elem.Cidr)
```
it seems to me that using fmt.Sprintf would perform better.

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	_, err = service.Id(vlanId).Mask("id").GetObject()
+	if err != nil {
+		return fmt.Errorf("Error deleting Network Vlan: %s", err)
+	}
```
Don't think we need to check if the vlan existed separately. I think the next call to the api below will tell you that already. If it existed, this results in two api calls, rather than just one.

minsikl replied:
Agree. Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	}
+	if billingItem.Id == nil {
+		return nil
+	}
```
should add a comment here based on our discussion

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+
+	if !success {
+		return fmt.Errorf("SoftLayer reported an unsuccessful cancellation")
+	}
```
Not sure. I think we should ignore success? and just return `err` back.

minsikl replied:
Not sure. But I couldn't find any cases in my test. I deleted that logic.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+		return false, err
+	}
+
+	return true, nil
```
here, you can do this at line 283: `return err == nil, err`

riethm replied:
Does this logic work?  i.e., If Terraform receives a non-nil error from `...Exists`, does it even consider the first (`bool`) return value, or would the error disqualify the `exists` test?

Thinking it also doesn't handle non-404 errors (500, network, etc.)

renier replied:
_helper/schema/resource.go_ says it checks the error first:

``` go
                exists, err := r.Exists(data, meta)
                if err != nil {
                        return s, err
                }
                if !exists {
                        return nil, nil
                }
```

you are saying we should check for 404 if we have an error?

``` go
if err != nil && err.(sl.Error).StatusCode == 404 {
        return false, nil
}

return err == nil, err
```

?

riethm replied:
Right.  I think that would work.

minsikl replied:
Applied

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+			vlans, err := services.GetAccountService(sess).
+				Filter(filter.Build(
+					filter.Path("networkVlans.billingItem.orderItem.order.id").
+						Eq(strconv.Itoa(orderId)))).
```
``` go
Filter(
        filter.Path("networkVlans.billingItem.orderItem.order.id").
        Eq(strconv.Itoa(orderId)).
        Build()
)
```

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	} else {
+		return &datatypes.Container_Product_Order_Network_Vlan{},
+			fmt.Errorf("datacenter name is empty.")
+	}
```
Would rewrite it as:

``` go
if datacenter == "" {
        return &datatypes.Container_Product_Order_Network_Vlan{},
                     fmt.Errorf("datacenter name is empty.")
}

dc, err := location.GetDatacenterByName(sess, datacenter, "id")
if err != nil {
    return &datatypes.Container_Product_Order_Network_Vlan{}, err
}
```

Then would not have to pre-declare dc and err above.

minsikl replied:
Applied.

renier commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	}
+
+	return &productOrderContainer, nil
+}
```
@riethm Would this function make a good candidate for adding as a helper to softlayer-go?

riethm replied:
@renier Seems like there is potential here for a helper.  For consideration: would the proposed helper just return the order container, or place the order (`OrderNetworkVlan()`) and return a receipt?

riethm commented on docs/softlayer_network_vlan.md:
```diff
@@ -0,0 +1,55 @@
+#### `softlayer_network_vlan`
```
Just want to put it out there - Are we sure we want to name this `softlayer_network_vlan` instead of perhaps shortening to just `softlayer_vlan` ?

renier replied:
👍

minsikl replied:
`softlayer_vlan` is better I think.

minsikl replied:
So change the resource name to `softlayer_vlan`?

minsikl replied:
Updated.

riethm commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+			"primary_router_hostname": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
```
Not sure I understand the implication of a property being declared both `Computed` and `Optional`.  You can always change a property's value programmatically, whether it is declared as `Computed` or not.  So this seems to be effectively the same as just making the property `Optional` only (`Computed` has no additional effect)?

minsikl replied:
Actually, I couldn't find other examples which use both Computed and Optional, but here is a test result of `Optional` property only.

_vlan.tf_

```
resource "softlayer_network_vlan" "test_vlan" {
   name = "test_vlan2"
   datacenter = "lon02"
   type = "PUBLIC"
   primary_subnet_size = 8
}
```

_terraform_ _show_ (after _terraform_ _apply_)

```
softlayer_network_vlan.test_vlan:
  id = 1369721
  child_resource_count = 0
  datacenter = lon02
  name = test_vlan2
  primary_router_hostname = fcr01a.lon02
  primary_subnet_size = 8
  softlayer_managed = false
  subnets.# = 1
  subnets.1384167489.subnet = 159.8.147.208/29
  subnets.1384167489.subnet_type = ADDITIONAL_PRIMARY
  type = PUBLIC
  vlan_number = 1255
```

_terraform_ _plan_

```
softlayer_network_vlan.test_vlan
primary_router_hostname: "fcr01a.lon02" => ""
```

With current codes, terraform tries to delete primary_router_hostname whenever I execute `terraform apply`

riethm replied:
I see.  That makes sense.  And because it's `ForceNew: true`, we never have to deal with the case where the config is updated.  Nice.

riethm commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	}
+
+	if len(router) > 0 {
+		rt, err = hardware.GetRouterByName(sess, router, "id")
```
It took me a few minutes to understand why the `GetRouterByName` helper was placed into a `hardware` package.  I do see why now, but still wonder if anyone would think to look there.  Maybe we should consider an alias function in another package (network?)  Just a thought - nothing to address for this PR.

minsikl replied:
Right. At first, I thought that Hardware or Hardware_Router services would provide those functions. However, GetRouterByName does not use any Hareware_\* services. Only a return type of GetRouterByName is `Hardware`. Except Router, other resources such as SoftLayer_Network_Pod and SoftLayer_Network_Backbone are using Network_\* datatypes. So their helper functions will be located in network package in the future. So network package can be good place for GetRouterByName.

riethm commented on softlayer/resource_softlayer_network_vlan.go:
```diff
@@ -0,0 +1,419 @@
+	} else {
+		return &datatypes.Container_Product_Order_Network_Vlan{},
+			fmt.Errorf("datacenter name is empty.")
+	}
```
👍

minsikl committed:
Update multiple logics in network_vlan

- Create a constant for vlan mask.
- Add a vlan type validataion fucntion to a schema.
- Use sl.Get to get "name" field.
- Use fmt.Sprintf to build a subnet string.
- Add comment in Delete function
- Remove VLAN object checking logic in Delete function
- Remove success flag from Delete function
- Simplify return logic in Exist function
- Update Filter in findNetworkVlanByOrderId
- buildVlanProductOrderContainer Refactoring

minsikl committed:
Change softlayer_network_vlan to softlayer_vlan
